src/main/java/org/testng/internal/annotations/IgnoreListener.java

```diff
@@ -0,0 +1,71 @@
+package org.testng.internal.annotations;
+
+import org.testng.annotations.ITestAnnotation;
+import org.testng.annotations.Ignore;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Method;
+
+public class IgnoreListener implements IAnnotationTransformer {
+
+    @Override
+    public void transform(ITestAnnotation annotation, Class testClass, Constructor testConstructor, Method testMethod) {
+        if (!annotation.getEnabled()) {
+            return;
+        }
+        Class<?> typedTestClass = testClass;
+        if (testMethod != null) {
+            ignoreTest(annotation, testMethod.getAnnotation(Ignore.class));
+            typedTestClass = testMethod.getDeclaringClass();
+        }
+        if (typedTestClass != null) {
+            ignoreTest(annotation, typedTestClass.getAnnotation(Ignore.class));
+            Package testPackage = typedTestClass.getPackage();
+            if (testPackage != null) {
+                ignoreTest(annotation, findAnnotation(testPackage));
+            }
+        }
+    }
+
+    private static void ignoreTest(ITestAnnotation annotation, Ignore ignore) {
+        if (ignore != null) {
+            annotation.setEnabled(false);
+            if (!ignore.value().isEmpty()) {
+                String ignoredDescription;
+                if (annotation.getDescription() == null || annotation.getDescription().isEmpty()) {
+                    ignoredDescription = ignore.value();
+                } else {
+                    ignoredDescription = ignore.value() + ": " + annotation.getDescription();
+                }
+                annotation.setDescription(ignoredDescription);
+            }
+        }
+    }
+
+    private static Ignore findAnnotation(Package testPackage) {
+        if (testPackage == null) {
+            return null;
+        }
+        Ignore result = testPackage.getAnnotation(Ignore.class);
+        if (result != null) {
+            return result;
+        }
+        String[] parts = testPackage.getName().split("\\.");
+        String parentPackageName = joinParent(".", parts);
+        if (parentPackageName.isEmpty()) {
+            return null;
+        }
+        return findAnnotation(Package.getPackage(parentPackageName));
+    }
+
+    private static String joinParent(String delimiter, String[] parts) {
```

Comment:
Can we move this method into the `org.testng.util.Strings` class ?

Reply:
Sure!

Reply:
done